### PR TITLE
fix(redo_courslet): 

### PR DESCRIPTION
**Pull request by alexbojko**
User can redo courslet, adds some tests to test this functionality

Fix issue #259 

@z4y4ts please review

### CODE REVIEW

File: mysite/chat/views.py [hunk 1]
```diff
@@ -100,14 +100,26 @@ def get(self, request, enroll_key):
...
+                user=request.user,
+                state__isnull=False
+            ).first()
         if not chat and enroll_key:
             chat = Chat(
                 user=request.user,
                 enroll_code=enroll_code,
                 instructor=courseUnit.course.addedBy
             )
-            chat.save(request)
```
Comment (alexbojko) [inline mysite/chat/views.py]: chat has no custom method save. that's why I deleted request

File: mysite/chat/views.py [hunk 2]
```diff
@@ -100,14 +100,26 @@ def get(self, request, enroll_key):
...
+                Chat,
+                enroll_code=enroll_code,
+                user=request.user,
+                id=chat_id
+            )
+        else:
+            chat = Chat.objects.filter(
+                enroll_code=enroll_code,
+                user=request.user,
+                state__isnull=False
```
Comment (alexbojko) [inline mysite/chat/views.py]: added state__isnull=False to get only not finished chats

File: mysite/lms/tests.py [hunk 3]
```diff
@@ -81,3 +84,130 @@ def test_course_view_negative(self):
...
+            lesson=self.lesson,
+            addedBy=self.user,
+            treeID=self.lesson.id,
+            order=0
+        )
+        self.unit_lesson.save()
+        self.user = User.objects.create_user(username='admin', password='admin')
+        call_command('fsm_deploy')
+
+    def test_courslet_history_tab(self):
```
Comment (alexbojko) [inline mysite/lms/tests.py]: this test works fine

File: mysite/lms/tests.py [hunk 4]
```diff
@@ -81,3 +84,130 @@ def test_course_view_negative(self):
...
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.context['courslet_history']), 0)
+        self.assertIsNotNone(list(response.context['courslets']))
+
+        self.assertEqual(response.status_code, 200)
+
+        chats_count_1 = Chat.objects.all().count()
+
+        response = self.client.get(
+            reverse('chat:chat_enroll', kwargs={'enroll_key': self.enroll.enrollCode})
```
Comment (alexbojko) [inline mysite/lms/tests.py]: create new chat - but this is ok now

File: mysite/lms/tests.py [hunk 5]
```diff
@@ -81,3 +84,130 @@ def test_course_view_negative(self):
...
+        )
+        self.assertEqual(response.context['chat'].id, Chat.objects.all().first().id)
+        self.assertEqual(response.status_code, 200)
+        chats_count_2 = Chat.objects.count()
+
+        self.assertNotEqual(chats_count_2, chats_count_1)
+
+        response = self.client.get(
+            reverse('chat:chat_enroll', kwargs={'enroll_key': self.enroll.enrollCode})
```
Comment (alexbojko) [inline mysite/lms/tests.py]: create new chat again, that's not fine though

File: mysite/lms/tests.py [hunk 6]
```diff
@@ -81,3 +84,130 @@ def test_course_view_negative(self):
...
+        )
+        chats_count_3 = Chat.objects.count()
+
+        response = self.client.get(
+            reverse('chat:chat_enroll', kwargs={'enroll_key': self.enroll.enrollCode})
```
Comment (alexbojko) [inline mysite/lms/tests.py]: create new chat again

**Comment by maxsocl** [comment]
@alexbojko first of all you have to rebase you code on latest master.

**Comment by maxsocl** [comment]
@alexbojko I've got an 404 error clicking on history of Courselet.

screen shot 2017-01-19 at 10 47 51 pm


screen shot 2017-01-19 at 10 47 44 pm

**Pull request merged**